File: servers/java/coweb-server/src/main/java/org/coweb/ModeratorLateJoinHandler.java
package org.coweb;

import java.util.Map;
import java.util.HashMap;
import java.util.logging.Logger;

import org.cometd.bayeux.Message;
import org.cometd.bayeux.server.ServerSession;

import org.eclipse.jetty.util.ajax.JSON;

public class ModeratorLateJoinHandler extends LateJoinHandler {
	private static final Logger log = Logger
			.getLogger(ModeratorLateJoinHandler.class.getName());


	public ModeratorLateJoinHandler(SessionHandler sessionHandler,
			Map<String, Object> config) {
		super(sessionHandler, config);	
	}

	@Override
	public void onClientJoin(ServerSession client, Message message) {
		log.info("ModeratorLateJoinHandler::onClientJoin *************");
		int siteId = this.getSiteForClient(client);

		if (siteId == -1) {
			siteId = this.addSiteForClient(client);
		}

		log.info("siteId = " + siteId);
		Map<Integer, String> roster = this.getRosterList(client);

		// Construct message: 1) app data, 2) engine state.
		Map<String, Object> collabs = this.sessionModerator.getLateJoinState();
		Object[] data = new Object[collabs.size() + 1];

		// 1) App data.
		int cnt = 0;
		for (Map.Entry<String, Object> collab: collabs.entrySet()) {
			HashMap<String, Object> hm = new HashMap<String, Object>();
			hm.put("topic", "coweb.state.set." + collab.getKey());
			hm.put("value", collab.getValue());
			data[cnt++] = hm;
		}

		// 2) Engine state.
		HashMap<String, Object> engState = new HashMap<String, Object>();
		engState.put("topic", "coweb.engine.state");
		engState.put("value", this.sessionHandler.getEngineState());
		data[cnt++] = engState;

		/* 3) Paused sync buffer. Since the moderator never pauses itself from processing sync events, we can safely
		  ignore sending this (we would be sending an empty array otherwise). */
		/* HashMap<String, Object> pauseState = new HashMap<String, Object>();
		pauseState.put("topic", "coweb.pause.state");
		pauseState.put("value", new Object[0]);
		data[cnt++] = pauseState; */

		if (this.updaters.isEmpty())
			this.addUpdater(client, false);
		else
			this.addUpdater(client, true);

		client.batch(new BatchUpdateMessage(client, siteId, roster, data, true));
	}

	public void onUpdaterSendState(ServerSession client, Message message) {
		// should never get here.
		return;
	}

	@Override
	protected void removeUpdater(ServerSession client) {
		this.removeSiteForClient(client);
		this.updaters.remove(client.getId());
		this.sendRosterUnavailable(client);
	}

}

File: servers/python/coweb/session/moderator_late_join_handler.py

from .late_join_handler import LateJoinHandler
import logging

log = logging.getLogger('coweb.session')

class ModeratorLateJoinHandler(LateJoinHandler):

    def __init__(self, session):
        super(ModeratorLateJoinHandler, self).__init__(session)
        self._moderator = session._moderator

    # Returns true iff this is the first client in the session.
    def onClientJoin(self, client):
        '''Overrides to have the moderator immediatelty send late join state'''
        clientId = client.clientId

        try:
            # get site id already assigned
            siteid = self.getSiteForClient(client)
        except ValueError:
            # get new site id
            siteid = self.addSiteForClient(client)

        # get the roster list to return before adding anyone
        roster = self._getRosterList(client)

        data = []
        collabs = self._moderator.getLateJoinState()
        for key in collabs:
            data.append({
                "topic" : "coweb.state.set." + key,
                "value" : collabs[key]
            })

        data.append({
            "topic" : "coweb.engine.state",
            "value" : self._session._opengine.getEngineState()
        })

        client.add_message({
            'channel':'/service/session/join/siteid',
            'data': siteid
        })
        client.add_message({
            'channel':'/service/session/join/roster',
            'data': roster
        })
        client.add_message({
            'channel':'/service/session/join/state',
            'data': data
        })

        if 0 == self.getUpdaterCount():
            self.addUpdater(client, False)
            return True
        else:
            self.addUpdater(client, True)
            return False

